Update bundled libffi to 3.3 and build without documentation

YorickPeterse opened the pull request:
This pull request makes two changes:

1. It updates the bundled libffi to the latest 3.3 version
2. It disables the building of documentation when building libffi, which removes the dependency on texinfo.

yorickpeterse commented:
Apparently this requires the use of `master`, as this option is not available in the latest release.

An RC is available per https://github.com/libffi/libffi/issues/296#issuecomment-377921049, but I'm not sure if using this is the best idea.

yorickpeterse commented:
An alternative would be to apply the following patch when building: https://patch-diff.githubusercontent.com/raw/libffi/libffi/pull/190.patch

This however means we need to require `patch`, instead of `texinfo`, which isn't much better.

@tov What do you think? Would it be acceptable to ship the RC, or would you prefer somehow applying the patch? It seems that a stable libffi release might still be a bit off.

tov commented:
I'm not sure. Given that this library is pre-stable, I don't inherently have a problem with shipping the RC, but maybe it's not worth it just to get rid of texinfo as a dependency? Or is the texinfo dependency a problem?

yorickpeterse commented:
@tov It's not necessarily a problem in that it prevents the use of libffi-sys-rs at this time, but it's an additional dependency to deal with, possible with different names per platform. 

What I think we can do is the following:

1. Prepare a new release of libffi-sys-rs/etc, so the fixes for Windows are available
1. Ask the libffi maintainers about the status of a new release. Based on how long a release would take, we can then either keep things as-is or somehow match the `configure` script.

I'll ask the libffi maintainers about the release.

tov commented:
Okay, so you want me to release what's on master right now?

yorickpeterse commented:
@tov Yes, that would be much appreciated! :smile:

tov commented:
@YorickPeterse Released as 0.6.3.

yorickpeterse commented:
@tov Thanks! :confetti_ball:

tov commented:
~It looks like `--disable-docs` is now part of the libffi release, which means that it’s probably time to merge this PR.~

Nope.

tov commented:
> I'll ask the libffi maintainers about the release.

Did you find anything out?

I was thinking that it would be easy to patch using `sed` instead of `patch`.

yorickpeterse commented:
@tov Since I last checked, no real progress has been made towards a new libffi release. Re-reading the issue, the last update from the maintainers was in April 2018. Based on this  I have my doubts about there being a new libffi release any time soon.

If we're going down the right of replacing strings, using a Rust "script" for this would be best so we don't add another dependency to the list (e.g. `sed`).

yorickpeterse commented:
@tov It has only taken a little over a year, but I rebased this on `master` and updated it for the latest libffi version (included in this PR). What do you think of these changes?
